base/compiler/abstractinterpretation.jl

```diff
@@ -900,26 +900,32 @@ function abstract_eval(@nospecialize(e), vtypes::VarTable, sv::InferenceState)
         t = instanceof_tfunc(abstract_eval(e.args[1], vtypes, sv))[1]
         if isconcretetype(t) && !t.mutable
             args = Vector{Any}(undef, length(e.args)-1)
-            isconst = true
+            ats = Vector{Any}(undef, length(e.args)-1)
+            anyconst = false
+            allconst = true
             for i = 2:length(e.args)
                 at = abstract_eval(e.args[i], vtypes, sv)
+                ats[i-1] = at
                 if at === Bottom
                     t = Bottom
-                    isconst = false
+                    allconst = anyconst = false
                     break
                 elseif at isa Const
                     if !(at.val isa fieldtype(t, i - 1))
                         t = Bottom
-                        isconst = false
+                        allconst = anyconst = false
                         break
                     end
+                    anyconst = true
```

Comment:
This should be set with the same `anyinfo` logic computation as in the `tuple_tfunc` (`# TODO ignore singleton Const` and following lines). It's admittedly less common to run into those cases in a struct than in a tuple, but no point in asking it to do a lot of extra work that we already know for certain does not improve the result.